### PR TITLE
Update on match modes for WMS-T layers static temporal range

### PR DESCRIPTION
**Pull request by Samweli**
Added two match modes in the time slice option in the static temporal range group found in raster layer temporal properties.

The "Previous Closest To Start Range" set the start of temporal range to the previous closest date time in the available temporal capabilities only if the requested temporal range is not found in the 
available temporal capabilities, while the "Previous Closest To End Range" mode does the same but using the end of the requested temporal range.

Update_on_match_modes

### CODE REVIEW

**Comment by timlinux** [comment]
Cc @tomkralidis

File: src/gui/raster/qgsrasterlayerproperties.cpp [hunk 1]
```diff
@@ -1290,6 +1281,8 @@ void QgsRasterLayerProperties::setSourceStaticTimeState()
     mFetchModeComboBox->addItem( tr( "Use Whole Temporal Range" ), QgsRasterDataProviderTemporalCapabilities::MatchUsingWholeRange );
     mFetchModeComboBox->addItem( tr( "Match to Start of Range" ), QgsRasterDataProviderTemporalCapabilities::MatchExactUsingStartOfRange );
     mFetchModeComboBox->addItem( tr( "Match to End of Range" ), QgsRasterDataProviderTemporalCapabilities::MatchExactUsingEndOfRange );
+    mFetchModeComboBox->addItem( tr( "Previous Closest to Start of Range" ), QgsRasterDataProviderTemporalCapabilities::FindClosestMatchToStartOfRange );
+    mFetchModeComboBox->addItem( tr( "Previous Closest to End of Range" ), QgsRasterDataProviderTemporalCapabilities::FindClosestMatchToEndOfRange );
```
Comment (nyalldawson) [inline src/gui/raster/qgsrasterlayerproperties.cpp:1285]:
```suggestion
    mFetchModeComboBox->addItem( tr( "Closest Match to Start of Range" ), QgsRasterDataProviderTemporalCapabilities::FindClosestMatchToStartOfRange );
    mFetchModeComboBox->addItem( tr( "Closest Match to End of Range" ), QgsRasterDataProviderTemporalCapabilities::FindClosestMatchToEndOfRange );
```

I find the "previous" wording confusing here

File: src/providers/wms/qgswmscapabilities.cpp [hunk 2]
```diff
@@ -334,6 +335,37 @@ QDateTime QgsWmsSettings::addTime( QDateTime dateTime, QgsWmstResolution resolut
   return resultDateTime;
 }
 
+QDateTime QgsWmsSettings::findLeastClosestDateTime( QDateTime dateTime )
+{
+  QDateTime closest = dateTime;
+  long long min = LLONG_MAX;
```
Comment (nyalldawson) [inline src/providers/wms/qgswmscapabilities.cpp]:
```suggestion
  long long min = std::numeric_limits<long long>::max();
```

File: src/providers/wms/qgswmscapabilities.h [hunk 3]
```diff
@@ -785,6 +785,14 @@ class QgsWmsSettings
 
     QDateTime addTime( QDateTime dateTime, QgsWmstResolution resolution );
 
+    /**
+     * Finds the least closest datetime from list of available datetimes
+     * with the given \a dateTime.
+     *
+     * Returns the passed \a dateTime if it is found in the available datetimes.
+     */
+    QDateTime findLeastClosestDateTime( QDateTime dateTime );
```
Comment (nyalldawson) [inline src/providers/wms/qgswmscapabilities.h]:
```suggestion
    QDateTime findLeastClosestDateTime( QDateTime dateTime ) const;
```

**Review by nyalldawson — COMMENTED** [review]
Nice addition!

**Review by nyalldawson — APPROVED** [review]

**Comment by elpaso** [comment]
@Samweli do you have any idea from where `| QGIS Server |`  in the dialog title comes from? It doesn't seem right to me.

File: src/providers/wms/qgswmscapabilities.cpp [hunk 4]
```diff
@@ -416,10 +448,10 @@ QgsWmstResolution QgsWmsSettings::parseWmstResolution( QString item )
 
 QDateTime QgsWmsSettings::parseWmstDateTimes( QString item )
 {
-  // Standard item will have YYYY-MM-DDThh:mm:ss.SSSZ
+  // Standard item will have YYYY-MM-DDTHH:mm:ss.SSSZ
   //  format a Qt::ISODateWithMs
 
-  QString format = "YYYY-MM-DDThh:mm:ss.SSSZ";
+  QString format = "YYYY-MM-DDTHH:mm:ss.SSSZ";
```
Comment (elpaso) [inline src/providers/wms/qgswmscapabilities.cpp:454]: QStringLiteral

**Comment by Samweli** [comment]
> @Samweli do you have any idea from where `| QGIS Server |` in the dialog title comes from? It doesn't seem right to me.

Hi @elpaso, I can't seem to find the `| QGIS Server |`.

**Comment by elpaso** [comment]
@Samweli I've filed bug:  https://github.com/qgis/QGIS/issues/35510

**Comment by Samweli** [comment]
> @Samweli I've filed bug: #35510

Okay, thanks @elpaso

**Pull request merged**